### PR TITLE
[core] Optimize value_accuracy_to_buf to avoid snprintf

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Replaces `snprintf("%.*f")` in `value_accuracy_to_buf` with integer-based formatting for finite float values. The `snprintf` float formatting machinery is extremely heavy — this change makes the function **14.3x faster**.

**Approach:**
- Scale the float value to an integer via `rint(double(value) * multiplier)`
- Split into integer and fractional parts using integer division
- Write digits directly to the buffer
- Uses `rint()` with double cast to match snprintf's banker's rounding exactly
- Falls back to `snprintf` only for NaN/Inf

**Benchmark (2000 inner iterations, host platform):**

| | Time (ns) | Items/s |
|---|---|---|
| Before | 92,961 | 21.5M/s |
| After | 6,484 | 308.5M/s |
| **Speedup** | **14.3x** | |

Also adds C++ unit tests (`test_value_accuracy.cpp`) that verify output matches `snprintf` for a wide range of values including negative, zero, large values, rounding edge cases, NaN, and Inf.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — [policy](https://developers.esphome.io/contributing/code/#what-constitutes-a-c-breaking-change)
- [ ] Developer breaking change (an API change that could break external components) — [policy](https://developers.esphome.io/contributing/code/#what-is-considered-public-c-api)
- [ ] Undocumented C++ API change (removal or change of undocumented public methods that lambda users may depend on) — [policy](https://developers.esphome.io/contributing/code/#c-user-expectations)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040/RP2350
- [ ] BK72xx
- [ ] RTL87xx
- [ ] LN882x
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# N/A - internal optimization only
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/core/helpers.cpp [hunk 1]
```diff
@@ -543,28 +543,52 @@ std::string value_accuracy_to_string(float value, int8_t accuracy_decimals) {
...
+  uint32_t mult = small_pow10(accuracy_decimals);
+  // Cast to double for the multiply to match snprintf's rounding precision.
+  // float*int loses bits at exact-half boundaries (e.g. 23.45f*10 = 234.5 in float,
+  // but snprintf sees 234.500007... via double promotion and rounds differently).
+  uint32_t scaled = static_cast<uint32_t>(lrint(static_cast<double>(value) * mult));
+  p = uint32_to_str_unchecked(p, scaled / mult);
+  if (accuracy_decimals > 0) {
+    *p++ = '.';
+    p = frac_to_str_unchecked(p, scaled % mult, mult / 10);
+  }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/helpers.cpp]: Fixed — switched to `llrint` (returns `long long`, so 32-bit `long` targets no longer truncate) and added an explicit range check in the caller: values whose scaled magnitude wouldn't fit in `uint32_t` (|value| ≥ UINT32_MAX/mult, ≈ 4.29e6 at 3 decimals) now fall through to `snprintf` instead of silently wrapping.

File: esphome/core/helpers.h [hunk 2]
```diff
@@ -1295,6 +1295,24 @@ inline char *int8_to_str(char *buf, int8_t val) {
   return buf;
 }
 
+/// Append a separator char and a string to a buffer, respecting remaining space.
+/// Returns pointer past last char written (null terminator is written).
+inline char *buf_append_sep_str(char *buf, size_t remaining, char separator, const char *str, size_t str_len) {
+  if (remaining < 2) {
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/helpers.h:1311]: Updated the docstring to make the early-return-without-terminator contract explicit, so future callers aren't misled.

File: tests/components/core/test_value_accuracy.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
+#include <gtest/gtest.h>
+#include <cmath>
+#include <cstdio>
+#include <cstring>
+#include <span>
+
+#include "esphome/core/helpers.h"
+
+namespace esphome::testing {
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/components/core/test_value_accuracy.cpp]: Fixed — renamed to `namespace esphome::core::testing` to match `test_helpers.cpp` and the other core test files.

File: tests/components/core/test_value_accuracy.cpp [hunk 4]
```diff
@@ -0,0 +1,151 @@
+#include <gtest/gtest.h>
+#include <cmath>
+#include <cstdio>
+#include <cstring>
+#include <span>
```
[the remark this answers is not in the document]
Comment (bdraco) [reply tests/components/core/test_value_accuracy.cpp:5]: Added `#include <string>` explicitly.

File: esphome/core/helpers.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/core/helpers.h:1311]: Took the suggested fix instead: now always null-terminates when `remaining >= 1`, so callers always get a valid C string regardless of path. Docstring updated to match.

**Review by jesserockz — APPROVED** [review]

**Comment by bdraco** [comment]
thanks
